mecha.zig

```diff
@@ -147,35 +240,49 @@ pub fn manyN(
     comptime n: usize,
     comptime options: ManyNOptions,
 ) Parser([n]ParserResult(@TypeOf(parser))) {
-    const Array = [n]ParserResult(@TypeOf(parser));
+    const T = @TypeOf(parser);
+    const Array = [n]ParserResult(T);
     const Res = Result(Array);
-    return .{ .parse = struct {
-        fn parse(allocator: mem.Allocator, str: []const u8) Error!Res {
-            var rem = str;
-            var res: Array = undefined;
-            for (&res, 0..) |*value, i| {
-                if (i != 0)
-                    rem = (try options.separator.parse(allocator, rem)).rest;
-
-                const r = try parser.parse(allocator, rem);
-                rem = r.rest;
-                value.* = r.value;
+    return .{
+        .parse = struct {
+            fn parse(allocator: mem.Allocator, str: []const u8) Error!Res {
+                var rem = str;
+                var res: Array = undefined;
+                var pos: usize = 0;
+                for (&res, 0..) |*value, i| {
+                    if (i != 0) {
+                        const sep = try options.separator.parse(allocator, rem);
+                        switch (sep) {
+                            .ok => |ok| {
+                                rem = ok.rest;
+                                pos += ok.span.len;
+                            },
+                            .err => return Res.failWith(sep, pos),
```

Comment:
I don't see the `failWith` function adding much. I'd just do this everywhere instead and remove it:

```suggestion
                            .err => |e| return Res.fail(e.loc, e.pos + pos),
```

Reply:
A type error was the reason for this change. I’ll try your suggestion and update this conversation.